File: Solutions/128. Longest Consecutive Sequence/128. Longest Consecutive Sequence.py
#HashSet implementation
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        nums=set(nums)
        best=0
        for x in nums:
            if x-1 not in nums:
                y=x+1
                while y in nums:
                    y+=1
                best=max(best,y-x)
        return best
        
        
#Sort
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0
        nums.sort()
        longseq,currseq=1,1
        for i in range(1,len(nums)):
            if nums[i]!=nums[i-1]:
                if nums[i]==nums[i-1]+1:
                    currseq+=1
                else:
                    longseq=max(longseq,currseq)
                    currseq=1
        return max(longseq,currseq)
 
 
#Union Find
class Solution:
    def find(self,ds:List[int],i:int)->int:
        return i if ds[i]<0 else self.find(ds,ds[i])
    def longestConsecutive(self, nums: List[int]) -> int:
        seq=0
        if len(nums)==0:
            return 0
        d={}
        ds=[-1]*len(nums)
        for i in range(len(nums)):
            n=nums[i]
            if n in d:
                continue
            d[n]=i
            if n-1 in d:
                j=self.find(ds,d[n])
                k=self.find(ds,d[n-1])
                ds[j]+=ds[k]
                ds[k]=j
            if n+1 in d:
                j=self.find(ds,d[n])
                k=self.find(ds,d[n+1])
                ds[j]+=ds[k]
                ds[k]=j
            seq=max(seq,abs(ds[self.find(ds,d[n])]))
        return seq

File: Solutions/128. Longest Consecutive Sequence/128. Longest Consecutive Sequence.java
// HashSet implementation
class Solution {
    public int longestConsecutive(int[] nums) {
        Integer[] num=new Integer[nums.length];
        for(int i=0;i<num.length;++i)
            num[i]=Integer.valueOf(nums[i]);
        Set<Integer> s = new HashSet<>(Arrays.asList(num)); 
        int longseq=0;
        for(var n:s){
            if(!s.contains(n-1)){
                int currseq=0;
                while(s.contains(n++))
                    ++currseq;
                longseq=Math.max(longseq,currseq);
            }
        }
        return longseq;
    }
}


//Sort,in java this is faster than above one
class Solution {
    public int longestConsecutive(int[] nums) {
        if(nums.length==0)
            return 0;
        Arrays.sort(nums);
        int currseq=1,longseq=1;
        for(int i=1;i<nums.length;++i){
            if(nums[i]!=nums[i-1]){
                if(nums[i]==nums[i-1]+1)
                    ++currseq;
                else{
                    longseq=Math.max(currseq,longseq);
                    currseq=1;
                }
            }
        }
        return Math.max(longseq,currseq);
    }
}


//Union Find
class Solution {
    private int find(int[] ds,int i){
        if(ds[i]<0)
            return i;
        ds[i]=find(ds,ds[i]);
        return ds[i];
    }
    public int longestConsecutive(int[] nums) {
        if(nums.length==0)
            return 0;
        int seq=0;
        int[] ds=new int[nums.length];
        Arrays.fill(ds,-1);
        Map<Integer,Integer> m=new HashMap<>();
        for(int i=0;i<nums.length;++i){
            int n=nums[i];
            if(m.containsKey(n))
                continue;
            m.put(n,i);
            if(m.containsKey(n-1)){
                //union
                int j=find(ds,m.get(n));
                int k=find(ds,m.get(n-1));
                ds[j]+=ds[k];
                ds[k]=j;
            }
            if(m.containsKey(n+1)){
                //union
                int j=find(ds,m.get(n));
                int k=find(ds,m.get(n+1));
                ds[j]+=ds[k];
                ds[k]=j;
            }
            seq=Math.max(seq,Math.abs(ds[find(ds,m.get(n))]));
        }
        return seq;
    }
}
